### PR TITLE
Relax only cg

### PR DESCRIPTION
**Pull request by eric-c-cyr**
I want ti videy stuff

### CODE REVIEW

File: torchbraid/odenet_apps.py [hunk 1]
```diff
@@ -141,6 +141,9 @@ def updateParallelWeights(self):
   def run(self,x):
     # turn on derivative path (as requried)
     self.use_deriv = self.training
+    
+    # instead of doing runBraid, can execute tests
+    #self.testBraid(x)
```
Comment (eric-c-cyr) [inline torchbraid/odenet_apps.py:146]: Thanks for adding this, didn't know about it.

File: torchbraid/torchbraid_app.pyx [hunk 2]
```diff
@@ -161,6 +162,7 @@ class BraidApp:
     braid_SetNRelax(core,0,0) # set F relax on fine grid
     braid_SetCFactor(core,-1,self.cfactor) # -1 implies chage on all levels
     braid_SetAbsTol(core,self.abs_tol)
+    #braid_SetCRelaxWt(core, -1, 1.2)   # Turn on weighted relaxation, probably want to add command line argument
```
Comment (eric-c-cyr) [inline torchbraid/torchbraid_app.pyx:165]: What is weighted relaxation?

File: torchbraid/torchbraid_app.pyx [hunk 3]
```diff
@@ -313,6 +365,10 @@ class BraidApp:
       core = (<PyBraid_Core> self.py_core).getCore()
       braid_SetPrintLevel(core,self.print_level)
 
+  def setMinCoarse(self, mc):
+    core = (<PyBraid_Core> self.py_core).getCore()
+    braid_SetMinCoarse(core, mc)
```
Comment (eric-c-cyr) [inline torchbraid/torchbraid_app.pyx:370]: What is the mc argument here? I have my own calculation I've been using, but is braid smart about this as well?

File: torchbraid/torchbraid_app.pyx [hunk 4]
```diff
@@ -440,4 +504,120 @@ class BraidApp:
...
+
+      # Get the coarse vector
+      _braid_UGetVectorRef(coarse_core, 0, coarse_pt, &coarse_bv)
+      assert(<unsigned int>(coarse_bv)!=0)
+ 
+      # Copy coarse to fine at C-pt
+      # --> ERIC please check how I do this deep copy from coarse_bv --> fine_bv 
+      py_fine_bv = <object> fine_bv.userVector
+      py_coarse_bv = <object> coarse_bv.userVector
+      py_coarse_bv_clone = BraidVector(torch.clone(py_coarse_bv.tensor_data_), 0)
```
Comment (eric-c-cyr) [inline torchbraid/torchbraid_app.pyx:559]: We might do `py_coarse_bv.tensor_data_.detach().clone()` here. The reason is that clone has semantics with regard to automatic differentiation. The call to detach separates the tensor from the AD graph, and then. allows a clean clone.

File: torchbraid/torchbraid_app.pyx [hunk 5]
```diff
@@ -440,4 +504,120 @@ class BraidApp:
...
+
+      # Get the coarse vector
+      _braid_UGetVectorRef(coarse_core, 0, coarse_pt, &coarse_bv)
+      assert(<unsigned int>(coarse_bv)!=0)
+ 
+      # Copy fine to coarse  
+      # --> ERIC please check how I do this deep copy from fine_bv --> coarse_bv 
+      py_fine_bv = <object> fine_bv.userVector
+      py_coarse_bv = <object> coarse_bv.userVector
+      py_fine_bv_clone = BraidVector(torch.clone(py_fine_bv.tensor_data_), 0)
```
Comment (eric-c-cyr) [inline torchbraid/torchbraid_app.pyx:619]: same comment as above respect to detach.

**Comment by eric-c-cyr** [comment]
@jbschroder i think this is fine in my mind based on what you asked in the comments. The only thing I would prefer is to use a `detach().clone()` instead so there are no hidden AD surprises (which wouldn't survive MPI anyway).

File: torchbraid/odenet_apps.py (hunk 1 above)
Comment (jbschroder) [reply torchbraid/odenet_apps.py:146]: I added these for David -- they aren't terribly useful, but they do ensure some basic functionality.

File: torchbraid/torchbraid_app.pyx (hunk 2 above)
Comment (jbschroder) [reply torchbraid/torchbraid_app.pyx:165]: This carries out weighted-Jacobi inside FCF.  It basically weight's the C-relaxation step, and turns one full FC-sweep into a weighted block-Jacobi relaxation, where the block size is the coarsening factor.

It can improve convergence (modestly), but it also allows us to smear solution values between XBraid solves and XBraid iterations.

File: torchbraid/torchbraid_app.pyx (hunk 3 above)
Comment (jbschroder) [reply torchbraid/torchbraid_app.pyx:370]: Braid will stop coarsening in time, if the next coarser grid will be smaller than mc.  So, if mc = 5, and your current coarse grid has 8 time points, then braid would stop, because 
8 / m < 5

File: torchbraid/torchbraid_app.pyx (hunk 4 above)
Comment (jbschroder) [reply torchbraid/torchbraid_app.pyx:559]: Done

File: torchbraid/torchbraid_app.pyx (hunk 5 above)
Comment (jbschroder) [reply torchbraid/torchbraid_app.pyx]: I just realized that "tensor_data_" is a tuple, so I have to modify the above code a little bit.  

But, then I realized that there is a braid_vector clone() function that I could use, that does this, too.  The only difference is that this clone function also copies "any weight tensors" and copies "layer information".  I don't have a good sense of what this other information is -- and if using that clone (in braid_vector.py) is a good idea.   

@eric-c-cyr  Let me know what you think

File: torchbraid/torchbraid_app.pyx (hunk 3 above)
Comment (eric-c-cyr) [reply torchbraid/torchbraid_app.pyx:370]: We should definitely pass this up to the layer-parallel level.

File: torchbraid/torchbraid_app.pyx (hunk 5 above)
Comment (eric-c-cyr) [reply torchbraid/torchbraid_app.pyx]: I think that probably makes the most sense. Just use the BraidVecotr.clone.

**Comment by eric-c-cyr** [comment]
@jbschroder I'm going to spend some time on this tomorrow. Is there a good place to start?

**Comment by jbschroder** [comment]
Hey Eric,

Hmmmm... There are some new options added to the LayerParallel solver
object (mostly to set Braid options like num levels, and num forward and
backward relax).  These might be the first things to pull in?  Then, the
examples would come next, as they rely on these new options?

I'll be around today writing a HW and working on my tenure slides, but I
can hop on for a quick Zoom chat here and there, as that would probably
make things easier.  Just shoot me a text or email if you want to do this.

Cheers,

Jacob
